File: java/pocjavaservice1/src/main/java/com/coichedid/pocjavaservice1/client/ServiceClient.java
package com.coichedid.pocjavaservice1.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.client.RestTemplate;

public class ServiceClient {
  private final RestTemplate restTemplate;

  private String gatewayHost;

  private String gatewayPort;

  private String gateway;
  private String nextService;

  private Logger logger = LoggerFactory.getLogger("com.coichedid.pocjavaservice1.client.ServiceCliet");

  public ServiceClient(RestTemplate restTemplate, String service, String gatewayHost) {
    this.gatewayHost = gatewayHost;
    this.restTemplate = restTemplate;
    this.gateway = "http://" + this.gatewayHost;
    this.nextService = service;
  }

  public String sayHello(int hops) {
    //Get an client to next service with URL like
    //http://gateway:port/api/nextServiceUrl/hops
    String nextServiceUrl = this.gateway + this.nextService + hops;
    logger.info("Calling next service with " + nextServiceUrl);
    String response = this.restTemplate
      .getForObject(nextServiceUrl, String.class);
    return response;
  }
}

File: nodejs/pocRequestGenerator/src/client/ServiceClient.js
import request from 'request-promise-native';

class ServiceClient {
  /**
   *
   * @param  {String} service     service to be connected
   * @param  {String} gatewayHost host of gateway
   * @return {ServiceClient}             new service client
   */
  constructor(gatewayHost) {
    this.gatewayHost = gatewayHost;
    this.gateway = `http://${this.gatewayHost}`;

    this.options = {
      method:'GET', //hardcoded we only have "gettable" services for this PoC
      uri:'',
      json:true
    }
  }

  /**
   * Call api of nextService url
   * @param  {number} hops Current hops executed
   * @return {String}      recursive response of this hop and next;
   */
  sayHello(nextService, hops) {
    return new Promise( (resolve, reject) => {
      this.options.uri = `${this.gateway}${nextService}${hops}`;
      request(this.options)
        .then(result => {
          resolve(result);
        })
        .catch( reason => {
          reject(reason);
        })
    });
  }
}

export default ServiceClient;
